### PR TITLE
Added :func:`~.override_animation` decorator

### LINKED ISSUE #1336: Add a decorator to `override_animation` in Mobject
## Description of proposed feature

As per [the discussion](https://github.com/ManimCommunity/manim/pull/1225#issuecomment-812216240) resulting in #1225 being closed, there's interest in providing Mobject specific animations to override Animation.

## How can the new feature be used?

Per @behackl suggestion, the decorator would be used like:

```python
    @override_animation(Create)
    def _custom_create(self, *args, **kwargs):
        # build a custom create animation
        return my_custom_animation
```
to override the way the class calls the animation (in this case Create).

## Additional comments

https://github.com/ManimCommunity/manim/pull/1225#issuecomment-812216240

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Changelog / Overview

Added `@override_animation(Animation)` decorator.

## Motivation
Closes #1336 

## Explanation for Changes
Have a look at the [docs](https://manimce--1422.org.readthedocs.build/en/1422/reference/manim.animation.override_animation.html?). If that leaves questions I should improve the docs :smile: 

## Personal Checklist
- [x] Basic Functionality
- [x] Allow using the default animation instead of overrides
- [x] Error if multiple methods override the same Animation
- [ ] Change docs of override method accordingly
- [ ] Change docs of overridden Animation accordingly (maybe)
- [x] Add Tests

Feel free to comment your ideas :)

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [x] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [x] The changelog entry is completed if necessary
- [x] Newly added functions/classes either have a docstring or are private
- [x] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by AntonBallmaier** [comment]
I have noticed a problem: When overriding the same animation in subclasses (and subsubclasses...) you would expect the inheritance behavior. But since overridden animation methods might be named differently, the typical inheritance doesn't work. I thought about it for quite a while and come to the conclusion that basically I would need to recreate pythons inheritance system. Therefor I suggest raising an error if inherited methods overriding the same animation aren't namend equally (and therefor actually overridden from an inheritance stand point).

**Comment by AntonBallmaier** [comment]
I would like to add two things to the docs:
 - For Methods overriding Animations I want to add a note that they are overriding that Animation.
 - For Animations I want to add a List of Mobjects overriding it. This might be useful when since if this isn't included in the docs someone might wonder why some Animations don't behave like they would expect them to. 
I do however need help with sphinx to add these. I would need a template for how such an entry could be formatted in markdown. Maybe someone wants to help? :upside_down_face:

File: manim/animation/override_animation.py
Comment (kilacoda) [inline manim/animation/override_animation.py:87]: shouldn't this be `animation_class` as per PEP8?

Comment (AntonBallmaier) [reply manim/animation/override_animation.py]: Yes, it's fixed

**Review by MrMallIronmaker — COMMENTED** [review]
Thanks for taking this feature on! I think it will be important especially as we build support for plugins.

So if I understand correctly, the overall architecture is:
- the Animation keeps track of the map from classes to override methods
- at "setup time," the program recursively iterates through all subclasses of Mobject and looks for functions with the decorator method
- at dispatch time

(BTW this feature-level overview for developers should be somewhere that isn't included with readthedocs, but still in the file, ideally.)

The primary question I have is whether a decorator is the right way to do this - @behackl are you set on a decorator for this ([c.f.](https://github.com/ManimCommunity/manim/pull/1225#issuecomment-812216240))?

It seems to me that here, a decorator is not really functioning as an extra layer to the function but rather a specifications or configuration file, mapping an animation class to one of the mobject's methods. Rather than the animation keeping track of the overrides, perhaps the animation should ask the Mobject if it has an override for a given animation - and it can go up the inheritance chain if need be.

There are a couple options - it could be a class dictionary mapping animation classes to methods, or it could be a Mobject method that takes in a class and either spits out the animation or None. With the dictionary example:

```py
class Animation:
    def __new__(
        cls,
        mobject: Optional[Mobject] = None,
        *args,
        use_default: bool = False,
        **kwargs,
    ):
        if (
            not use_default
            and mobject
            and cls in mobject.animation_overrides
        ):
            # make the overridden anim
            return anim
        return super().__new__(cls)
    # snip

class MyMobject(Mobject):
    def create(self, run_time):
        # snip
    animation_overrides = {
        manim.Create: create
    }
```

First off, would this also get the job done, or are there design concerns you've found that fit the decorator process better?

Second, would this be better than 'trying to reimplement the Python inheritance system'?

**Comment by AntonBallmaier** [comment]
> So if I understand correctly, the overall architecture is:
> 
>     * the Animation keeps track of the map from classes to override methods
> 
>     * at "setup time," the program recursively iterates through all subclasses of Mobject and looks for functions with the decorator method
> 
>     * at dispatch time

Yeah, that's the idea.

> (BTW this feature-level overview for developers should be somewhere that isn't included with readthedocs, but still in the file, ideally.)

I'll add that as soon as we have decided on one concept.

> 
> The primary question I have is whether a decorator is the right way to do this - @behackl are you set on a decorator for this ([c.f.](https://github.com/ManimCommunity/manim/pull/1225#issuecomment-812216240))?
> 
> It seems to me that here, a decorator is not really functioning as an extra layer to the function but rather a specifications or configuration file, mapping an animation class to one of the mobject's methods. Rather than the animation keeping track of the overrides, perhaps the animation should ask the Mobject if it has an override for a given animation - and it can go up the inheritance chain if need be.

This is true. A decorator would be a great way to do this if there was a way to access the class a method is defined in at compile time, but there doesn't seem to be one. Therefor I had to do the setup-function workaround.

In the end it doesn't really matter where the mapping is saved - imo Animation and Mobject classes are equally fitting places. 

> There are a couple options - it could be a class dictionary mapping animation classes to methods, or it could be a Mobject method that takes in a class and either spits out the animation or None. With the dictionary example:
> 
> ```python
> ...
> ```
> 
> First off, would this also get the job done, or are there design concerns you've found that fit the decorator process better?

I think this is a valid approach. It might not be as intuitive syntactically, but in turn we could get rid of the setup function. It kind of feels like a config dictionary 2.0  : / 

> Second, would this be better than 'trying to reimplement the Python inheritance system'?
Definitely. However, I managed to solve the inheritance issue with my version reliably as well.

**Comment by MrMallIronmaker** [comment]
I was hoping behackl would leave a comment. I'm on the fence unfortunately - it doesn't feel like it should be a decorator in my opinion, but I don't feel confident enough to go either way.

Do you know how this would work with plugins? Would setup be run before any plugins?

**Comment by friedkeenan** [comment]
I think it being a decorator is fine. A decorator is something that performs an action on a function/whatever, whether that be wrapping it/adding a layer on top of it, or registering it for something. The already existing `override_animate` functions as a decorator, and I believe that's good.

As for "if there was a way to access the class a method is defined in at compile time, but there doesn't seem to be one", this is correct. You cannot get the class a decorator is being used in at the time the decorator is being executed (especially as at that point, the class isn't instantiated yet). However, you *can* work around this reasonably well by storing some information in the decorated function's attributes and then later looping over the class's methods (such as on [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__)) and acting on the tagged functions appropriately.

I'll also try to review this soon

**Comment by AntonBallmaier** [comment]
> However, you _can_ work around this reasonably well by storing some information in the decorated function's attributes and then later looping over the class's methods (such as on [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__)) and acting on the tagged functions appropriately.

If I understand you correctly this is basically what I did, right?

**Comment by friedkeenan** [comment]
> If I understand you correctly this is basically what I did, right?

No, I don't believe so. It's the same idea roughly but there's a bit of nuance in how they're different. Importantly, I don't believe how you've done it would work for user-created mobject subclasses as all the logic is done when manim is imported

**Comment by AntonBallmaier** [comment]
> > If I understand you correctly this is basically what I did, right?
> 
> No, I don't believe so. It's the same idea roughly but there's a bit of nuance in how they're different. Importantly, I don't believe how you've done it would work for user-created mobject subclasses as all the logic is done when manim is imported

When creating subclasses you have to call the setup function afterwards with my method, right. There is a way to fix that? I would really appreciate halp with this since I am quite new to python and don't know how to do that.

**Comment by MrMallIronmaker** [comment]
I'm not 100% sure how `__init_subclass__` works (i.e., I can't write up the pseudocode for it), but it looks pretty reasonable for this use case. In short, when a new class subclassing Mobject is defined, it can look through for any method decorators methods and they can be added to the map of animation redirects.

That change would address my concerns about the architecture.

You might even be able to remove the initial `setup_animation_overriding` call because the work would be done as each subclass of Mobject is defined.

I'm happy to help how I can in explaining [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) but I just learned of this two hours ago :)

**Comment by friedkeenan** [comment]
`__init_subclass__` is called when a subclass (i.e. of `Mobject`) is initialized. If you need a more concrete reference, I use it in my draft PR for mobject groups that I should work on more (#1353)

**Comment by AntonBallmaier** [comment]
I moved to __init_subclass__ and it works as intended :) Thanks for the tip

**Comment by friedkeenan** [comment]
Requesting my own review because I do have things to request changes for, but do not atm have time for it (short term business), just in case someone's trigger finger on the merge button is a little itchy and merges before I can review

File: manim/animation/override_animation.py
Comment (naveen521kk) [inline manim/animation/override_animation.py:9]: use relative imports
```suggestion
from ..mobject import Mobject
from ..utils.exceptions import MultiAnimationOverrideException
```

File: manim/animation/override_animation.py
Comment (naveen521kk) [inline manim/animation/override_animation.py:16]: you don't need to use strings.
```suggestion
    mobject_class: Type(Mobject),
    overrides: Dict[Type[Animation], List[Dict[Type[Mobject], str]]],
```

Comment (naveen521kk) [reply manim/animation/override_animation.py:16]: Also, why you should use `Type`?

File: manim/animation/override_animation.py
Comment (naveen521kk) [inline manim/animation/override_animation.py:54]: it's better to simply raise the msg.
```suggestion
                raise MultiAnimationOverrideException(msg)
```

File: manim/animation/override_animation.py
Comment (naveen521kk) [inline manim/animation/override_animation.py:64]:
```suggestion
    animation_class: Type[Animation],
```

Comment (naveen521kk) [reply manim/animation/override_animation.py:64]: Also, like above is this correct?

File: manim/mobject/mobject.py
Comment (naveen521kk) [inline manim/mobject/mobject.py:80]:
```suggestion
    animation_overrides: Dict[Type[Animation], List[Dict[Type[Mobject], str]]] = {}
```

**Review by naveen521kk — CHANGES_REQUESTED** [review]

File: manim/animation/override_animation.py
Comment (AntonBallmaier) [reply manim/animation/override_animation.py:16]: > Also, why you should use `Type`?

Type[X] means that a (sub)class of X is expected, not an instance.

Comment (AntonBallmaier) [reply manim/animation/override_animation.py:16]: > you don't need to use strings.

As far as I know I do. Typing can be done with or without quotation marks - preferably without. However, that requires to import the class into the module, which in this case would cause circular import issues.

Comment (naveen521kk) [reply manim/animation/override_animation.py:15]: > this case would cause circular import issues.

You should anyway import them even if it is a string for linters like mypy to know about them. To avoid circular import you can import it conditionally using https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING.

File: manim/mobject/mobject.py
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:80]: Is there a workaround for Mobject as well? I cannot use it inside the class... I changed it to `Dict[Type[Animation], List[Dict[Type["Mobject"], str]]]`

File: manim/animation/override_animation.py
Comment (AntonBallmaier) [reply manim/animation/override_animation.py:15]: ...that creates a new problem: Now I get the error `NameError: name 'Animation' is not defined` when testing

Comment (AntonBallmaier) [reply manim/animation/override_animation.py:15]: I changed it to only enquote Animation, not the whole type.

Comment (AntonBallmaier) [reply manim/animation/override_animation.py:15]: As it states in the linked page: Only local variables can be typed using this method.

Comment (naveen521kk) [reply manim/animation/override_animation.py]: > Now I get the error NameError: name 'Animation' is not defined when testing

You should import Animation conditionally and use it as a string. But since we have dropped support for py3.6 you can import `from __future__ import annotations` at the top of the file and then use then without quotes.

Comment (naveen521kk) [reply manim/animation/override_animation.py]: Though you ignore it and use string and it's completely fine.

Comment (AntonBallmaier) [reply manim/animation/override_animation.py]: ok ^^

File: manim/animation/override_animation.py
Comment (friedkeenan) [inline manim/animation/override_animation.py:54]: This should just be changed to `raise MultiAnimationOverrideException(...)` with no `msg` variable

File: manim/animation/override_animation.py
Comment (friedkeenan) [inline manim/animation/override_animation.py:63]: This should be in animation.py imo. With this and the earlier logic moved to be under `Mobject`, the whole override_animation.py probably wouldn't be necessary anymore.

File: manim/animation/override_animation.py
Comment (friedkeenan) [inline manim/animation/override_animation.py:59]: Don't think this is necessary, as I think the override functions for animations shouldn't be called directly and instead the overridden animation should be called.
```suggestion
```

In any case the commented out code should definitely not be here

File: manim/animation/override_animation.py
Comment (friedkeenan) [inline manim/animation/override_animation.py:89]: With my thoughts on how override functions shouldn't be called directly and the animations they're overriding called instead, this function should be private.
```suggestion
            def _fade_in_override(self, **kwargs):
```

File: manim/mobject/logo.py
Comment (friedkeenan) [inline manim/mobject/logo.py:113]: I'll note here that I'm not sure whether this method should be private or not. For backwards compatibility this is probably fine as is, and for harmony with the `expand` method, but I wonder if maybe ultimately users shouldn't call this method directly.

File: tests/test_override_animation.py
Comment (friedkeenan) [inline tests/test_override_animation.py:81]:
```suggestion
```

**Review by friedkeenan — COMMENTED** [review]
First pass of a review, mainly requesting some logic shifting

File: manim/animation/animation.py
Comment (AntonBallmaier) [reply manim/animation/animation.py:43]: Moving the logic to `Mobject` is a good idea, but I don't think this should be a classmethod. Usage and function itself would be cleaner as normal method imo.

Comment (friedkeenan) [reply manim/animation/animation.py:43]: What I envision is something like
```py
override_func = mobject.animation_override_for(cls) # Will call the classmethod
return override_func(mobject, *args, **kwargs)
```
in `Animation.__new__`. I don't think it makes that much sense as a normal method if the overrides are coming from the class, but you know maybe animation overrides could be specific to mobject instances and not their classes, and then there wouldn't be a need for `__init_subclass__` even, the logic could just live in `Mobject.__init__`

File: manim/mobject/logo.py
Comment (AntonBallmaier) [reply manim/mobject/logo.py:113]: I think it should be `create` for the same reasons.

File: manim/animation/override_animation.py
Comment (AntonBallmaier) [reply manim/animation/override_animation.py]: Yes, that was wrong. I fixed that.

**Comment by AntonBallmaier** [comment]
> First pass of a review, mainly requesting some logic shifting

Thanks a lot for the great suggestions, they make the code a lot more less convoluted. Time for the second pass, I guess :D

File: tests/test_override_animation.py
Comment (friedkeenan) [reply tests/test_override_animation.py:81]: I guess because it's defining a class. I disagree with black here, but whatever

File: docs/source/reference.rst
Comment (friedkeenan) [inline docs/source/reference.rst:136]:
```suggestion
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:87]: Don't think this should be here now that classes store their own overrides.
```suggestion
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:89]: I know this automatically gets converted to a classmethod by python, but I'd prefer being explicit about it nonetheless. Also best practice is to forward `**kwargs` on to the parent method, even if `object.__init_subclass__` just errors if it gets keyword arguments.
```suggestion
    @classmethod
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:124]: Nonblocking, but I'd prefer this to not start with `get_`, feels a tad unpythonic. `animation_override_for` would be something I'd prefer.

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:127]:
```suggestion

        return None
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:133]:
```suggestion
            # Ignore dunder methods
            if method_name.startswith("__"):
                continue

```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:141]:
```suggestion
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:140]: Nonblocking, but perhaps a `remove_animation_override` method could be added too.

File: manim/animation/animation.py
Comment (friedkeenan) [inline manim/animation/animation.py:303]:
```suggestion

    See Also
    --------
    :meth:`~.Mobject.add_animation_override`

```

**Review by friedkeenan — COMMENTED** [review]
Second pass, mainly style requests

File: manim/mobject/mobject.py
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:87]: In my opinion the annotation helps to understand the structure how the overrides are saved, but I probably can move it somewhere else.

Comment (friedkeenan) [reply manim/mobject/mobject.py:87]: A comment above `cls.animation_overrides = {}` (or even a type annotation there) seems appropriate to me

File: manim/mobject/mobject.py
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:127]: Ok, get seems to me like a good name, but I am new to python, so I'll simply trust you :D

File: manim/mobject/mobject.py
Comment (AntonBallmaier) [reply manim/mobject/mobject.py]: Yes, that's what I wanted to do as well

File: manim/mobject/mobject.py
Comment (AntonBallmaier) [reply manim/mobject/mobject.py:146]: At the moment I have not a lot of time to spare so I would rather not :)

File: manim/animation/animation.py
Comment (friedkeenan) [inline manim/animation/animation.py:31]: Nonlblocking but changing the check to
```suggestion
        if mobject is not None and use_override:
```
makes more sense to me

**Review by friedkeenan — APPROVED** [review]
One small suggestion, but I approve

File: manim/animation/animation.py
Comment (friedkeenan) [reply manim/animation/animation.py:31]: mm, that makes sense

**Comment by AntonBallmaier** [comment]
Should we wait for @behackl s review? This PR has been reviewed by 4 people already, and until now he hasn't interacted with this PR at all.

**Comment by MrMallIronmaker** [comment]
I requested the review from behackl so I just un-requested it.

I did request Naveen's re-review; it looks like his was a couple simple changes so I think it's fine to merge tomorrow if nothing else is raised.

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:127]: Nonblocking but it'd be cool if this could get docs

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:146]: Nonblocking again but it would be cool if this could get docs too

**Review by friedkeenan — COMMENTED** [review]
This doesn't change my approval, just some extra things

File: manim/mobject/mobject.py
Comment (naveen521kk) [inline manim/mobject/mobject.py:131]: Not returning anything means returning `None`, so I think this is redundant.

File: manim/mobject/mobject.py
Comment (naveen521kk) [reply manim/mobject/mobject.py:127]: also, typing.

**Review by naveen521kk — APPROVED** [review]
LGTM, some comments

File: manim/mobject/mobject.py
Comment (friedkeenan) [reply manim/mobject/mobject.py:131]: I think it's better to explicitly return `None` as to me, not returning anything means "it doesn't make sense to use the return value", but explicitly returning `None` signals that `None` is a valid, intentional return value that participates in the logic.

File: manim/mobject/mobject.py
Comment (naveen521kk) [inline manim/mobject/mobject.py:34]: relative imports
```suggestion
from ..animation.animation import override_animation
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:140]:
```suggestion
            The function returning the override animation or ``None`` if no such animation
```

File: manim/mobject/mobject.py
Comment (friedkeenan) [inline manim/mobject/mobject.py:150]:
```suggestion
        """Initializes animation overrides marked with the :func:`~.override_animation`
```

**Comment by behackl** [comment]
> Should we wait for @behackl s review? This PR has been reviewed by 4 people already, and until now he hasn't interacted with this PR at all.

Sorry, I've accumulated quite a backlog from being unavailable for a while; still working through that. I've briefly looked at the changes here -- LGTM overall, and particularly good job with the tests! This can certainly be merged as soon as everyone already involved here is happy. :-)

**Pull request merged**
